src/mod_mam_elasticsearch_arch.erl

```diff
@@ -0,0 +1,288 @@
+
+-spec build_search_query(map()) -> mongoose_elasticsearch:query().
+build_search_query(Params) ->
+    Filters = build_filters(Params),
+    TextSearchQuery = build_text_search_query(Params),
+    #{query =>
+      #{bool =>
+        #{must => TextSearchQuery,
+          filter => Filters}}}.
+
+-spec build_filters(map()) -> [map()].
+build_filters(Params) ->
+    Builders = [fun owner_filter/1,
+                fun with_jid_filter/1,
+                fun range_filter/1],
+    lists:flatmap(fun(F) -> F(Params) end, Builders).
+
+-spec owner_filter(map()) -> [map()].
+owner_filter(#{owner_jid := Owner}) ->
+    BinOwner = mod_mam_utils:bare_jid(Owner),
+    [#{term => #{owner => BinOwner}}].
+
+-spec with_jid_filter(map()) -> [map()].
+with_jid_filter(#{with_jid := #jid{} = WithJid}) ->
+    [#{term => #{remote => mod_mam_utils:bare_jid(WithJid)}}];
+with_jid_filter(_) ->
+    [].
+
+-spec range_filter(map()) -> [map()].
+range_filter(#{end_ts := End, start_ts := Start, borders := Borders, rsm := RSM}) ->
+    {StartId, EndId} = mod_mam_utils:calculate_msg_id_borders(RSM, Borders, Start, End),
+    Range1 = maybe_add_end_filter(EndId, #{}),
+    Range2 = maybe_add_start_filter(StartId, Range1),
+
+    case maps:size(Range2) of
+        0 ->
+            [];
+        _ ->
+            [#{range => #{mam_id => Range2}}]
+    end;
+range_filter(_) ->
+    [].
+
+-spec maybe_add_end_filter(undefined | mod_mam:message_id(), map()) -> map().
+maybe_add_end_filter(undefined, RangeMap) ->
+    RangeMap;
+maybe_add_end_filter(Value, RangeMap) ->
+    RangeMap#{lt => Value}.
+
+-spec maybe_add_start_filter(undefined | mod_mam:message_id(), map()) -> map().
+maybe_add_start_filter(undefined, RangeMap) ->
+    RangeMap;
+maybe_add_start_filter(Value, RangeMap) ->
+    RangeMap#{gt => Value}.
+
+-spec build_text_search_query(map()) -> map().
+build_text_search_query(#{search_text := SearchText}) when is_binary(SearchText) ->
+    #{simple_query_string => #{query => SearchText,
+                               fields => [<<"body">>],
+                               default_operator => <<"and">>}};
+build_text_search_query(_) ->
+    #{match_all => #{}}.
+
+-spec determine_sorting(map()) -> asc | desc.
+determine_sorting(#{rsm := #rsm_in{direction = before}}) ->
+    desc;
+determine_sorting(_) ->
+    asc.
+
+-spec maybe_add_from_constraint(mongoose_elasticsearch:query(), map()) ->
+    mongoose_elasticsearch:query().
+maybe_add_from_constraint(Query, #{rsm := #rsm_in{index = Offset}}) when is_integer(Offset) ->
+    Query#{from => Offset};
+maybe_add_from_constraint(Query, _) ->
+    Query.
+
+-spec search_result_to_mam_lookup_result(map(), map()) -> mod_mam:lookup_result().
+search_result_to_mam_lookup_result(Result, Params) ->
+    #{<<"hits">> :=
+      #{<<"hits">> := Hits,
+        <<"total">> := TotalCount}} = Result,
+
+    Messages = lists:sort(
+                 lists:map(fun hit_to_mam_message/1, Hits)),
+
+    case maps:get(is_simple, Params) of
```

Comment:
```erlang
+        <<"total">> := TotalCount}} = Result,
+
+    Messages = lists:sort(
+                 lists:map(fun hit_to_mam_message/1, Hits)),
+
+    case maps:get(is_simple, Params) of
```

becomes

```erlang
+        <<"total">> := TotalCount}} = Result,
+    Messages = lists:sort(lists:map(fun hit_to_mam_message/1, Hits)),
+    case maps:get(is_simple, Params) of
```

Reply:
```erlang
Messages = lists:sort(lists:map(fun hit_to_mam_message/1, Hits)),
```

I don't like it. It's a bunch of function calls stuffed together. Line break makes it more obvious what transformations we're applying.

Reply:
I see the point.
But we don't use new lines in our code, so it looks different, comparing to our other code.

Reply:
Also, not auto-code formatter friendly :)

Reply:
What kind of auto-code formatter? Do we have a set of guidelines for formatting the code?

Reply:
There are no auto-formatters yet. 
There is no code style yet.

There is ESL code style and https://github.com/inaka/erlang_guidelines which we don't follow.

Reply:
ok, can be kept like this. Until someone writes code style standards.


But error-logger calls should be patched :)

Reply:
Yup, I'll fix the log format.